File: bun-civitai-browser-master/src/civitai-api/v1/models/model-version.ts
import { type } from "arktype";
import {
  modelFileSchema,
  modelImageSchema,
} from "./shared-types";

/**
 * Model version stats schema for the model-version endpoint
 * Note: This endpoint has different stats structure (no thumbsDownCount)
 */
export const modelVersionEndpointStatsSchema = type({
  downloadCount: "number.integer",
  "ratingCount?": "number.integer",
  "rating?": "number",
  thumbsUpCount: "number.integer",
  // thumbsDownCount is not present in this endpoint
});
export type ModelVersionEndpointStats = typeof modelVersionEndpointStatsSchema.infer;

/**
 * Model version schema for the model-version endpoint
 * Represents a specific model version with detailed information
 */
export const modelVersionEndpointDataSchema = type({
  id: "number.integer",
  modelId: "number.integer",
  // index: "number.integer", // the position in modelId.modelVersions array (not present in this endpoint)
  name: "string",
  baseModel: "string",
  baseModelType: "string | null",
  publishedAt: "string.date.iso", // ISO 8601
  // availability: "'EarlyAccess' | 'Public'", // not present in this endpoint
  nsfwLevel: "number.integer",
  description: "string | null", // html doc strings
  "trainedWords?": "string[]",
  stats: modelVersionEndpointStatsSchema,
  // covered: 'boolean', // have cover image or not
  files: modelFileSchema.array(),
  images: modelImageSchema.array(),
});
export type ModelVersionEndpointData = typeof modelVersionEndpointDataSchema.infer;

File: lib/civitai_api/models/model_version.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'shared.dart';

part 'model_version.freezed.dart';
part 'model_version.g.dart';

/// A specific model version as returned by GET /model-versions/:id.
/// Note: stats structure differs from the models endpoint (no thumbsDownCount).
@freezed
abstract class ModelVersionEndpointData with _$ModelVersionEndpointData {
  const factory ModelVersionEndpointData({
    required int id,
    required int modelId,
    required String name,
    required String baseModel,
    String? baseModelType,
    required DateTime publishedAt,
    required int nsfwLevel,
    String? description,
    @Default([]) List<String> trainedWords,
    @Default(ModelVersionEndpointStats()) ModelVersionEndpointStats stats,
    @Default([]) List<ModelFile> files,
    @Default([]) List<ModelImage> images,
  }) = _ModelVersionEndpointData;

  factory ModelVersionEndpointData.fromJson(Map<String, dynamic> json) =>
      _$ModelVersionEndpointDataFromJson(json);
}
